tests/test_discriminated_union.py

```diff
@@ -1232,3 +1232,72 @@ class TestModel(BaseModel):
         'title': 'TestModel',
         'type': 'object',
     }
+
+
+def test_field_validator_wrapping_union() -> None:
+    """https://github.com/pydantic/pydantic/issues/6525"""
+
+    class CatModel(BaseModel):
+        name: Literal['kitty', 'cat']
+
+        @field_validator('name', mode='after')
+        def replace_name(cls, v: str) -> str:
+            return 'cat'
+
+    class SparrowModel(BaseModel):
+        name: Literal['sparrow', 'birdie']
+
+        @field_validator('name', mode='after')
+        def replace_name(cls, v: str) -> str:
+            return 'sparrow'
+
+    Animal = Annotated[
+        CatModel | SparrowModel,
+        Field(discriminator='name'),
+    ]
+
+    ta = TypeAdapter(Animal)
+
+    assert ta.validate_python({'name': 'kitty'}) == CatModel(name='cat')
+
+    with pytest.raises(ValidationError) as exc_info:
+        ta.validate_python({'name': 'rabbit'})
+
+    # insert_assert(exc_info.value.errors(include_url=False))
+    assert exc_info.value.errors(include_url=False) == [
+        {
+            'type': 'union_tag_invalid',
+            'loc': (),
+            'msg': "Input tag 'rabbit' found using 'name' does not match any of the expected tags: 'kitty', 'cat', 'sparrow', 'birdie'",  # noqa: E501
+            'input': {'name': 'rabbit'},
+            'ctx': {'discriminator': "'name'", 'tag': 'rabbit', 'expected_tags': "'kitty', 'cat', 'sparrow', 'birdie'"},
+        }
+    ]
+
+    # insert_assert(ta.json_schema())
+    assert ta.json_schema() == {
+        '$defs': {
+            'CatModel': {
+                'properties': {'name': {'enum': ['kitty', 'cat'], 'title': 'Name', 'type': 'string'}},
+                'required': ['name'],
+                'title': 'CatModel',
+                'type': 'object',
+            },
+            'SparrowModel': {
+                'properties': {'name': {'enum': ['sparrow', 'birdie'], 'title': 'Name', 'type': 'string'}},
+                'required': ['name'],
+                'title': 'SparrowModel',
+                'type': 'object',
+            },
+        },
+        'discriminator': {
+            'mapping': {
+                'birdie': '#/$defs/SparrowModel',
+                'cat': '#/$defs/CatModel',
+                'kitty': '#/$defs/CatModel',
+                'sparrow': '#/$defs/SparrowModel',
+            },
+            'propertyName': 'name',
+        },
+        'oneOf': [{'$ref': '#/$defs/CatModel'}, {'$ref': '#/$defs/SparrowModel'}],
+    }
```

Comment:
Thorough! But a bit tedious to write + replicate. Any way to simplify this part of the code?

If what we care about are the `$defs` then maybe:
```
    assert ta.json_schema() == {
        '$defs': {
            CatModel.json_schema(),
            SparrowModel.json_schema(),
        },
        'discriminator': {
            'mapping': {
                'birdie': '#/$defs/SparrowModel',
                'cat': '#/$defs/CatModel',
                'kitty': '#/$defs/CatModel',
                'sparrow': '#/$defs/SparrowModel',
            },
            'propertyName': 'name',
        },
        'oneOf': [{'$ref': '#/$defs/CatModel'}, {'$ref': '#/$defs/SparrowModel'}],
    }
```

Reply:
Indeed that'd be less verbose, but the `insert_assert` utility does snapshot testing so this is not hand written and doesn't need to be updated by hand 😀